<h1 style="color:blue">Data analysis | Netflix</h1>

Netflix is one of the most popular media and video streaming platforms. They have over 8000 movies or tv shows available on their platform, as of mid-2021, they have over 200M Subscribers globally. This tabular dataset consists of listings of all the movies and tv shows available on Netflix, along with details such as - cast, directors, ratings, release year, duration, etc.

##
<h2 style="color:blue">1. Understanding the data</h2>

In this section, I used pandas to explore my dataset, analyzing its characteristics, dimensions, and statistical properties.

    The section includes:
- summary statistics such as mean, median, mode, standard deviation, and quartiles, to understand the central tendency and distribution of the data for numerical columns;
- conclusions based on these summary statistics;
- summary statistics for categorical columns; and
- conclusions based on these statistics.

In [1]:
# Imported libraries necessary for the whole data journey.

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
print(pd.__version__)
print(np.__version__)
print(plt.matplotlib.__version__)
print(sns.__version__)

2.2.1
1.26.4
3.8.3
0.13.2


In [3]:
# Loaded the dataset into a dataframe.

netflix_df = pd.read_csv("/Users/milenko/My Drive (1307mile@gmail.com)/bootcamp/w3 project/data/netflix_titles.csv")

# Identified the dimensions of the dataset. It has 8807 rows and 12 columns.

netflix_df.shape

(8807, 12)

In [4]:
# Inspected the dataframe for cleanliness, and identified the content of the columns. The data appears to be clean.

netflix_df

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

##
<h2 style="color:blue">1.1. Data description</h2>

- "**show_id**" : a sequence of numbers in ascending order, with prefix "s".
- Prefix "s" may mean that the show ID starts with an "s", and a movie ID starts with "m"?

- "**type**" : content-categories, "Movie" and "TV Show".

- "**title**" : name of the content.

- "**director**" : full name of a person who directed the content.

- "**cast**" : full names of some of the stars of the content.

- "**country**" : country or countries where the content was produced.

- "**date_added**" : the date the content was published on Netflix.
Original format of the values in the column are "Month DD, YYYY".

- "**release_year**" : content release year.

- "**rating**" : content rating for the US market; provides information about the appropriate audience of the content.

- "**duration**" : timelength of the content, denoted in minute formats or number of seasons.

- "**listed_in**" : genre/s of the content.

- "**description**" : content-plot.

In [5]:
# Checked the data types of each column. All the types seem appropriate for the nature of the variable.

netflix_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [6]:
# Checked for the number of unique values for each column to determine which columns appear to be categorical.

netflix_df.nunique()

show_id         8807
type               2
title           8807
director        4528
cast            7692
country          748
date_added      1767
release_year      74
rating            17
duration         220
listed_in        514
description     8775
dtype: int64

In [7]:
# Divided the columns to categorical and numerical data,
# based on data preview, column-names and content, data type, and number of unique values per column.

numerical_data = ["release_year"]
categorical_data = ["show_id", "type", "title", "director", "cast", "country", "date_added", "rating", "duration", "listed_in", "description"]

##
<h2 style="color:blue">1.2. Numerical data</h2>

Here is statistical summary of the numerical data, which in this case is only 1 column, "release_year".<br>
Of the 8807 titles available on Netflix, average release year is 2014, give or take 9 years.<br>
The oldest title in the library is from 1925, and the most recent one is from 2021, the year of the data collection.

In [8]:
netflix_df[numerical_data].describe().round(0).astype(int)

release_year
count          8807
mean           2014
std               9
min            1925
25%            2013
50%            2017
75%            2019
max            2021

In [9]:
# Observed the count of unique values. As the statistical summary showed, 75% of the content was released between 2013 and 2019.

netflix_df["release_year"].value_counts()

release_year
2018    1147
2017    1032
2019    1030
2020     953
2016     902
        ... 
1959       1
1925       1
1961       1
1947       1
1966       1
Name: count, Length: 74, dtype: int64

In [10]:
# Observed unique values for "release_year". They look clean and in consistent format YYYY, ready for analysis.

unique_release_years = np.sort(netflix_df["release_year"].unique())
print(unique_release_years)

[1925 1942 1943 1944 1945 1946 1947 1954 1955 1956 1958 1959 1960 1961
 1962 1963 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975
 1976 1977 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989
 1990 1991 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003
 2004 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017
 2018 2019 2020 2021]


##
<h2 style="color:blue">1.3. Categorical data</h2>

Statistical summary of the categorical data shows that:
- **count-wise**, 8 columns are well-populated, almost 100%, while 4 columns are missing 10-20% values.
- **unique** count in some columns seems unexpectedly high, such as 748 countries, or 514 genres.
  
- **top** values are as follows:
- Movie is the dominant content **type**, with 6131 entries against 2676 TV Shows.
- Rajiv Chilaka and David Attenborough show as the top single entries in terms of **director**s and **cast**.
- United States is the **country** where relative majority of the content was produced.
- January 1, 2020 was the peak day when 109 movies or tv shows were **added**.
- 1/3 of the content is targetting and restricted to a mature audience, as per its **rating**.
- 20% of the content had **duration** of 1 Season.
- Most frequently the content was **listed_in** or as "Dramas, International Movies".

In [11]:
netflix_df[categorical_data].describe()

show_id   type                 title       director  \
count     8807   8807                  8807           6173   
unique    8807      2                  8807           4528   
top         s1  Movie  Dick Johnson Is Dead  Rajiv Chilaka   
freq         1   6131                     1             19   

                      cast        country       date_added rating  duration  \
count                 7982           7976             8797   8803      8804   
unique                7692            748             1767     17       220   
top     David Attenborough  United States  January 1, 2020  TV-MA  1 Season   
freq                    19           2818              109   3207      1793   

                           listed_in  \
count                           8807   
unique                           514   
top     Dramas, International Movies   
freq                             362   

                                              description  
count                                                8807  
unique                                               8775  
top     Paranormal activity at a lush, abandoned prope...  
freq                                                    4

In [12]:
# When observing columns such as "country", unique values count feels odd, because we don't have 748 countries in the world, right?
# Let's check the value counts to see what is causing this.

for column in categorical_data:
    value_counts = netflix_df[column].value_counts()
    print(f"Value counts for {column}:\n{value_counts}\n")

Value counts for show_id:
show_id
s1       1
s5875    1
s5869    1
s5870    1
s5871    1
        ..
s2931    1
s2930    1
s2929    1
s2928    1
s8807    1
Name: count, Length: 8807, dtype: int64

Value counts for type:
type
Movie      6131
TV Show    2676
Name: count, dtype: int64

Value counts for title:
title
Dick Johnson Is Dead                     1
Ip Man 2                                 1
Hannibal Buress: Comedy Camisado         1
Turbo FAST                               1
Masha's Tales                            1
                                        ..
Love for Sale 2                          1
ROAD TO ROMA                             1
Good Time                                1
Captain Underpants Epic Choice-o-Rama    1
Zubaan                                   1
Name: count, Length: 8807, dtype: int64

Value counts for director:
director
Rajiv Chilaka                     19
Raúl Campos, Jan Suter            18
Marcus Raboy                      16
Suhas Kadav               

The high count of unique values is caused by multiple entries per field. Let's explain it on the example of countries:
- content produced in the United States generates one unique value for the column "country";
- co-production between the US and, say, France, generates its own unique value "United States, France" for the column "country".
  
So, while half of the content is produced by only 4 countries, it's the co-productions, co-directed content, and other joint efforts that possibly inflate some columns, and this will need to be parsed.

In [13]:
# While looking for the cause of inflated number of unique values, I also spotted leading whitespace in some values.
# Hint: ', France, Algeria' in "country", and ' April 15, 2018' in "date_added". So, it's time to start cleaning this data.

for column in categorical_data:
    unique = np.sort(netflix_df[column].astype(str).unique())
    print(f"Unique values for {column}:\n{unique}\n")

Unique values for show_id:
['s1' 's10' 's100' ... 's997' 's998' 's999']

Unique values for type:
['Movie' 'TV Show']

Unique values for title:
['#Alive' '#AnneFrank - Parallel Stories' '#FriendButMarried' ... '마녀사냥'
 '반드시 잡는다' '최강전사 미니특공대 : 영웅의 탄생']

Unique values for director:
['A. L. Vijay' 'A. Raajdheep' 'A. Salaam' ... 'Óskar Thór Axelsson'
 'Ömer Faruk Sorak' 'Şenol Sönmez']

Unique values for cast:
["'Najite Dede, Jude Chukwuka, Taiwo Arimoro, Odenike Odetola, Funmi Eko, Keppy Ekpenyong"
 '4Minute, B1A4, BtoB, ELSIE, EXID, EXO, Got7, INFINITE, KARA, Shinee, Sistar, VIXX, Nine Muses, BTS, Secret, Topp Dogg'
 '50 Cent, Ryan Phillippe, Bruce Willis, Rory Markham, Jenna Dewan, Brett Granstaff, Randy Couture, Susie Abromeit, Ron Turner, James Remar'
 ...
 'Şahin Irmak, İrem Sak, Gonca Vuslateri, Emre Karayel, Duygu Yetiş, Onur Buldu, Salih Kalyon, Bilge Şen, Nilgün Belgün, Hakan Akın'
 'Şükrü Özyıldız, Aslı Enver, Şenay Gürler, Başak Parlak, Mahir Günşiray, Hakan Boyav, Hakan Gerçek, 

##
<h2 style="color:blue">2. Data cleaning and formatting</h2>

What happens in this section:
- cleaning data:
  - dealing with null values,
  - duplicates,
  - redundant or useless info, and
- formatting.

In [14]:
# Checked for null values. They are concentrated in 3 columns: "director", "cast", and "country".
# Filled the null values with "unknown" string.

netflix_df.isna().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [15]:
#netflix_df = netflix_df.fillna("unknown")
netflix_df

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

In [16]:
# Checked if any entry in the column "show_id" starts with a letter other than "s". There is no other prefix.

any_other_letter = netflix_df.show_id[netflix_df.show_id.str[0].str.lower() != 's']
print(any_other_letter)

Series([], Name: show_id, dtype: object)


In [17]:
# Removed the column "show_id". It's redundant with the default index.

netflix_df = 


SyntaxError: invalid syntax (3877044300.py, line 3)

Used many data cleaning techniques including handling null values, duplicates, dropping unnecessary columns, manipulating strings, and formatting data
• Addressed missing data properly, fully justified the imputation strategy

In [ ]:
# Changed "date_added" column to datatype more inappropriate for calculations, and corrected its mixed format.

netflix_df.date_added = pd.to_datetime(netflix_df.date_added, format="mixed")

netflix_df.info()
netflix_df

In [ ]:
clean whitespace

In [ ]:
comma delimiter for grouped values